### PR TITLE
[WIP] osdsdock refactoring

### LINKED ISSUE #957: Framwork:  load southbound storage driver via golang plugin
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
> /kind feature

**What happened**:
Currentlly , all southbound drivers are builtin code, when we run osdsdock process, all driver will load in memory nomatter we need it or not.also some drivers have many thirdparty dependency package (e.g. ceph driver need to depoly librados-dev and librbd-dev). This cause osdsdock more and more hard to depoly.

**What you expected to happen**:
golang has supported plugin in version 1.8 ,so we can make the driver plugable.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Hotpot(release/branch) version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by xxwjj**
**What this PR does / why we need it**:
1. load driver dynamicaly.
2. refactor the directory stuctrue.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fix #957 
**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: contrib/drivers/volume/huawei/oceanstor/replication.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ package oceanstor
 
 import (
 	"fmt"
+	"github.com/opensds/opensds/contrib/drivers/factory"
```
Comment (leonwanghui) [inline contrib/drivers/volume/huawei/oceanstor/replication.go]: Same here.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -23,9 +23,9 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/opensds/opensds/contrib/drivers/utils/constants"
```
Comment (leonwanghui) [inline pkg/controller/controller.go]: Same here.

File: pkg/dock/discovery/discovery.go [hunk 3]
```diff
@@ -21,6 +21,7 @@ package discovery
 
 import (
 	"fmt"
+	consts "github.com/opensds/opensds/contrib/drivers/utils/constants"
```
Comment (leonwanghui) [inline pkg/dock/discovery/discovery.go]: Same here.

File: pkg/dock/dock.go [hunk 4]
```diff
@@ -23,18 +23,20 @@ import (
...
+	"github.com/opensds/opensds/contrib/drivers/fileshare"
+	"github.com/opensds/opensds/contrib/drivers/utils/constants"
+	"github.com/opensds/opensds/contrib/drivers/volume"
 	c "github.com/opensds/opensds/pkg/context"
 	"github.com/opensds/opensds/pkg/db"
 	"github.com/opensds/opensds/pkg/dock/discovery"
 	"github.com/opensds/opensds/pkg/model"
 	pb "github.com/opensds/opensds/pkg/model/proto"
 	"google.golang.org/grpc"
+	"net"
```
Comment (leonwanghui) [inline pkg/dock/dock.go]: Same here.

File: Makefile [hunk 5]
```diff
@@ -38,7 +40,11 @@ prebuild:
 .PHONY: osdsdock osdslet osdsapiserver osdsctl docker test protoc goimports
 
 osdsdock:
-	go build -ldflags '-w -s' -o $(BUILD_DIR)/bin/osdsdock github.com/opensds/opensds/cmd/osdsdock
+	go build -tags $(OSDS_DOCK_TAGS) -ldflags '-w -s' -o $(BUILD_DIR)/bin/osdsdock github.com/opensds/opensds/cmd/osdsdock
+ifeq ($(strip $(OSDS_DOCK_TAGS)), driver_plugin)
+	make -f block_pulgin.mk
```
Comment (rhsakarpos) [inline Makefile]: please change pulgin to plugin

File: contrib/drivers/driver_factory_builtin.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	DriverNameManila:          manila.NewDriver,
+}
+
+var drivers = map[string]map[string]NewDriver{}
+
+func init() {
+	drivers[string(StorageTypeBlock)+DriverTypeProvision.String()] = VolumeDriverMap
+	drivers[string(StorageTypeBlock)+DriverTypeReplication.String()] = VolumeReplicationDriverMap
+	drivers[string(StorageTypeBlock)+DriverTypeMetric.String()] = VolumeMetricsDriverMap
+	drivers[string(StorageTypeFile)+DriverTypeProvision.String()] = FileShareDriverMap
```
Comment (rhsakarpos) [inline contrib/drivers/driver_factory_builtin.go]: DriverTypeProvision is repeated, please check

File: contrib/drivers/driver_factory_builtin.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func (d *DriverFactory) GetDriver(st constants.StorageType, dt constants.DriverType,
+	bp config.BackendProperties) (factory.Driver, error) {
+	driverMap, ok := drivers[string(st)+dt.String()]
+	if !ok {
+		return nil, fmt.Errorf("cann't find specified driver map, storageType:%s, driverType:%s", st, dt)
```
Comment (rhsakarpos) [inline contrib/drivers/driver_factory_builtin.go]: cann't -> cannot

File: contrib/drivers/driver_factory_plugin.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		libPaths = append(libPaths, path.Join(gopath, "src/github.com/opensds/opensds/build/out/lib"))
+	}
+	libPaths = append(libPaths, "/opt/opensds-hotpot-linux-amd64/lib")
+	libPaths = append(libPaths, "/usr/local/opensds/lib")
+
+	for _, libPath := range libPaths {
+		p := path.Join(libPath, string(storageType), driverName+".so")
+		if ok, _ := utils.PathExists(p); ok {
+			logs.Info("find %s driver so in path %s", driverName, p)
+			return p, nil
```
Comment (rhsakarpos) [inline contrib/drivers/driver_factory_plugin.go]: should the driver path(s) be cached once found and validated ?

File: contrib/drivers/driver_manager.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+type DriverWrapper struct {
+	// driver name
+	Name string
+	// Base Driver
+	factory.Driver
+	// whether driver initialized used by atomic
+	initialized int32
+}
+
+func (b *DriverWrapper) SetInitialized(value bool) {
```
Comment (rhsakarpos) [inline contrib/drivers/driver_manager.go]: is this thread safe ?

File: contrib/drivers/volume/huawei/oceanstor/oceanstor.go [hunk 10]
```diff
@@ -21,27 +21,33 @@ import (
...
 
 func (d *Driver) Setup() (err error) {
 	// Read huawei oceanstor config file
 	conf := &OceanStorConfig{}
 	d.conf = conf
-	path := config.CONF.OsdsDock.Backends.HuaweiOceanStorBlock.ConfigPath
-
-	if "" == path {
-		path = defaultConfPath
+	path := defaultConfPath
```
Comment (rhsakarpos) [inline contrib/drivers/volume/huawei/oceanstor/oceanstor.go]: this piece of code is repeated in all drivers...please add to base class/struct (DriverBase) or similar

**Comment by kumarashit** [comment]
Closing now. Will see a proper it properly next time

**Pull request closed without merging**